model/legacy-private/src/main/java/org/keycloak/models/dblock/DBLockGlobalLockProvider.java

```diff
@@ -47,14 +50,19 @@ private static DBLockProvider.Namespace stringToNamespace(String lockName) {
     }
 
     @Override
-    public GlobalLock acquire(String lockName, Duration timeToWaitForLock) {
+    public <V> V withLock(String lockName, Duration timeToWaitForLock, KeycloakSessionTaskWithResult<V> task) {
         Objects.requireNonNull(lockName, "lockName cannot be null");
 
         if (timeToWaitForLock != null) {
             LOG.debug("DBLockGlobalLockProvider does not support setting timeToWaitForLock per lock.");
         }
-        dbLockProvider.waitForLock(stringToNamespace(lockName));
-        return () -> releaseLock(lockName);
+
+        try {
+            dbLockProvider.waitForLock(stringToNamespace(lockName));
+            return task.run(session);
+        } finally {
+            releaseLock(lockName);
+        }
```

Comment:
This doesn't use the semantics described in the JavaDoc: 
* lock is not acquired in a separate transaction
* task is not run in a separate transaction
* lock is not released in a separate transaction

If this is not possible differently for the legacy store, this would at least require a comment.

Reply:
Good point, I think the locking and unlocking will be impossible to do in separate tx based on how `DBLockProvider` works, so I would prefer to add just a comment. However, the task probably should run in a separate tx. I tried to add it and I get the following exception. Any idea what can be causing it? Can the same happen when we will have some db update for the new store?

```
 java.lang.RuntimeException: Failed to update database
	at org.keycloak.quarkus.runtime.storage.legacy.liquibase.QuarkusJpaUpdaterProvider.update(QuarkusJpaUpdaterProvider.java:122)
	at org.keycloak.quarkus.runtime.storage.legacy.liquibase.QuarkusJpaUpdaterProvider.update(QuarkusJpaUpdaterProvider.java:85)
	at org.keycloak.quarkus.runtime.storage.legacy.database.LegacyJpaConnectionProviderFactory.lambda$update$1(LegacyJpaConnectionProviderFactory.java:298)
	at org.keycloak.models.utils.KeycloakModelUtils.runJobInTransactionWithResult(KeycloakModelUtils.java:268)
	at org.keycloak.models.dblock.DBLockGlobalLockProvider.withLock(DBLockGlobalLockProvider.java:67)
	at org.keycloak.models.locking.GlobalLockProvider.withLock(GlobalLockProvider.java:45)
	at org.keycloak.quarkus.runtime.storage.legacy.database.LegacyJpaConnectionProviderFactory.update(LegacyJpaConnectionProviderFactory.java:297)
	at org.keycloak.quarkus.runtime.storage.legacy.database.LegacyJpaConnectionProviderFactory.createOrUpdateSchema(LegacyJpaConnectionProviderFactory.java:265)
	at org.keycloak.quarkus.runtime.storage.legacy.database.LegacyJpaConnectionProviderFactory.postInit(LegacyJpaConnectionProviderFactory.java:131)
	at org.keycloak.quarkus.runtime.integration.QuarkusKeycloakSessionFactory.init(QuarkusKeycloakSessionFactory.java:104)
	at org.keycloak.quarkus.runtime.integration.jaxrs.QuarkusKeycloakApplication.startup(QuarkusKeycloakApplication.java:42)
	at org.keycloak.quarkus.runtime.integration.QuarkusLifecycleObserver.onStartupEvent(QuarkusLifecycleObserver.java:37)
	at org.keycloak.quarkus.runtime.integration.QuarkusLifecycleObserver_Observer_onStartupEvent_b0e82415b143738dc1f986a5fa4668e83d0a5dea.notify(Unknown Source)
	at io.quarkus.arc.impl.EventImpl$Notifier.notifyObservers(EventImpl.java:326)
	at io.quarkus.arc.impl.EventImpl$Notifier.notify(EventImpl.java:308)
	at io.quarkus.arc.impl.EventImpl.fire(EventImpl.java:76)
	at io.quarkus.arc.runtime.ArcRecorder.fireLifecycleEvent(ArcRecorder.java:131)
	at io.quarkus.arc.runtime.ArcRecorder.handleLifecycleEvents(ArcRecorder.java:100)
	at io.quarkus.deployment.steps.LifecycleEventsBuildStep$startupEvent1144526294.deploy_0(Unknown Source)
	at io.quarkus.deployment.steps.LifecycleEventsBuildStep$startupEvent1144526294.deploy(Unknown Source)
	at io.quarkus.runner.ApplicationImpl.doStart(Unknown Source)
	at io.quarkus.runtime.Application.start(Application.java:101)
	at io.quarkus.runtime.ApplicationLifecycleManager.run(ApplicationLifecycleManager.java:110)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:70)
	at org.keycloak.quarkus.runtime.KeycloakMain.start(KeycloakMain.java:103)
	at org.keycloak.quarkus.runtime.cli.command.AbstractStartCommand.run(AbstractStartCommand.java:37)
	at picocli.CommandLine.executeUserObject(CommandLine.java:1939)
	at picocli.CommandLine.access$1300(CommandLine.java:145)
	at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2358)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2352)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2314)
	at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2179)
	at picocli.CommandLine$RunLast.execute(CommandLine.java:2316)
	at picocli.CommandLine.execute(CommandLine.java:2078)
	at org.keycloak.quarkus.runtime.cli.Picocli.parseAndRun(Picocli.java:93)
	at org.keycloak.quarkus.runtime.KeycloakMain.main(KeycloakMain.java:89)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at io.quarkus.bootstrap.runner.QuarkusEntryPoint.doRun(QuarkusEntryPoint.java:61)
	at io.quarkus.bootstrap.runner.QuarkusEntryPoint.main(QuarkusEntryPoint.java:32)
Caused by: liquibase.exception.DatabaseException: java.sql.SQLException: Deferred enlistment not supported
	at liquibase.database.jvm.JdbcConnection.getDatabaseProductName(JdbcConnection.java:111)
	at liquibase.database.core.H2Database.isCorrectDatabaseImplementation(H2Database.java:204)
	at liquibase.database.DatabaseFactory.findCorrectDatabaseImplementation(DatabaseFactory.java:101)
	at org.keycloak.quarkus.runtime.storage.legacy.liquibase.QuarkusLiquibaseConnectionProvider.getLiquibase(QuarkusLiquibaseConnectionProvider.java:124)
	at org.keycloak.quarkus.runtime.storage.legacy.liquibase.QuarkusJpaUpdaterProvider.getLiquibaseForKeycloakUpdate(QuarkusJpaUpdaterProvider.java:306)
	at org.keycloak.quarkus.runtime.storage.legacy.liquibase.QuarkusJpaUpdaterProvider.update(QuarkusJpaUpdaterProvider.java:103)
	... 41 more
Caused by: java.sql.SQLException: Deferred enlistment not supported
	at io.agroal.pool.ConnectionHandler.verifyEnlistment(ConnectionHandler.java:391)
	at io.agroal.pool.wrapper.ConnectionWrapper.getMetaData(ConnectionWrapper.java:452)
	at liquibase.database.jvm.JdbcConnection.getDatabaseProductName(JdbcConnection.java:109)
	... 46 more
```